### PR TITLE
Notification Service for AppsApi

### PR DESCRIPTION
**Pull request by bradenshipley**
## Description of change

This PR adds a sidekiq job to `vets-api` to send notification emails for each user who has connected to or disconnected from an application in the Application Directory within the last hour. We do this by hitting the Okta systems log api and pulling all connection/disconnection events in a specific set of time and checking each event for if the application referenced is included in the Application Directory. If it is, we hit `.send_email` with all information in the `personalization_hash` that the referenced email template number requires.
Unit tests have been included where possible, however the VANotify Notification API can't be tested locally, so we need to add a flag we can check against to know if we are in the staging environment or not so that testing can be done in that environment. A PR has been made to add this flag to each environment in the devops repo under `ansible/deployments/config/vets-api/<env>-settings.local.yml.j2` with the value being `true` for staging and `false` for every other environment. Once testing is done in the staging environment, a PR will be made to remove references to the staging flag.

## Original issue(s)
See JIRA ticket here: https://vajira.max.gov/secure/RapidBoard.jspa?rapidView=777&view=detail&selectedIssue=API-3829
See devops PR here (closed and merged): https://github.com/department-of-veterans-affairs/devops/pull/8514

## Things to know about this PR

Dummy data has been added to settings.yml but nothing of consequence.

### CODE REVIEW

File: config/sidekiq_scheduler.yml
Comment (ericboehs) [inline config/sidekiq_scheduler.yml:181]: Can you tell me more about the removal of this job? Is this no longer needed? Can it be removed?

File: modules/apps_api/README.md
Comment (ericboehs) [inline modules/apps_api/README.md:3]: Does it? https://github.com/department-of-veterans-affairs/vets-api/blob/master/LICENSE.md

File: modules/apps_api/README.md
Comment (ericboehs) [inline modules/apps_api/README.md:21]:
```suggestion
For internal usage, there is a [CLI](https://github.com/department-of-veterans-affairs/app_directory_cli) available to allow for interacting with the `#create`, `#update`, and `#destroy` methods. For syntax and usage of that CLI, see the linked repo. Non-GET methods used in this CLI require an api_key for the environment you are trying to make changes in.
```

File: spec/support/vcr_cassettes/okta/connection_logs.yml
Comment (ericboehs) [inline spec/support/vcr_cassettes/okta/connection_logs.yml:36]: There's no secrets in these here cassettes? Right? RIGHT?

File: config/sidekiq_scheduler.yml
Comment (bradenshipley) [reply config/sidekiq_scheduler.yml:181]: This sidekiq job should have been removed in a previous PR (https://github.com/department-of-veterans-affairs/vets-api/pull/5862) that removed the `AppsApi::DirectoryLoader` class in a refactor. It slipped through the cracks and does nothing currently so its removal got added to this pr.

File: modules/apps_api/README.md
Comment (bradenshipley) [reply modules/apps_api/README.md:3]: This is an error that has been fixed in the latest commit.

File: modules/apps_api/README.md
Comment (bradenshipley) [reply modules/apps_api/README.md:21]: Ah whoops! Nice catch

Comment (bradenshipley) [reply modules/apps_api/README.md:21]: fixed in latest commit

File: spec/support/vcr_cassettes/okta/connection_logs.yml
Comment (bradenshipley) [reply spec/support/vcr_cassettes/okta/connection_logs.yml:36]: it's just the Public Key Information, and is in every other cassette on vets-api. So as far as I know it's fine.
Screen Shot 2021-02-23 at 3 55 52 PM

File: modules/apps_api/lib/apps_api/notification_service.rb
Comment (ffafara-tw) [inline modules/apps_api/lib/apps_api/notification_service.rb:14]: We have introduced changes to VaNotify service. It requires you to pass api key as argument to the constructor. Please take a look at: https://github.com/department-of-veterans-affairs/vets-api/pull/5962.

**Review by ffafara-tw — DISMISSED** [review]

File: modules/apps_api/lib/apps_api/notification_service.rb
Comment (bradenshipley) [reply modules/apps_api/lib/apps_api/notification_service.rb:14]: Ah thanks for posting this! I was just looking at the tests failing and feeling really confused for a second. I've made updates to the code to reference the api_key addition

Comment (ffafara-tw) [reply modules/apps_api/lib/apps_api/notification_service.rb]: You might want to use lighthouse api key.
Settings.vanotify.services.lighthouse.api_key

It might not be in config/settings.yml - @saman-moshafi-va was supposed to add it today.
However it should be there in devops repo for all real environments.

Comment (bradenshipley) [reply modules/apps_api/lib/apps_api/notification_service.rb]: Ah, I see. I've updated it to be correct.

File: modules/apps_api/app/workers/apps_api/fetch_connections.rb
Comment (ffafara-tw) [inline modules/apps_api/app/workers/apps_api/fetch_connections.rb:16]: This should be `Settings.vanotify.lighthouse.template_id.connection_template`

And I believe @saman-moshafi-va is changing the Settings yaml structure to have 'templates' instead of 'template_id'.
So it will be:
`Settings.vanotify.lighthouse.templates.connection_template`
after his changes.

Sorry for all this confusion. It all is happening all at once...

File: modules/apps_api/app/workers/apps_api/fetch_connections.rb
Comment (ffafara-tw) [inline modules/apps_api/app/workers/apps_api/fetch_connections.rb:21]: This should be `Settings.vanotify.lighthouse.template_id.connection_template`
or `Settings.vanotify.lighthouse.templates.connection_template`

@saman-moshafi-va where are you with the changes to settings yaml structure?

File: modules/apps_api/app/workers/apps_api/fetch_connections.rb
Comment (saman-moshafi-va) [reply modules/apps_api/app/workers/apps_api/fetch_connections.rb]: The entries for the lighthouse templates will be `Settings.vanotify.lighthouse.template_id.connection_template` and `Settings.vanotify.lighthouse.template_id.disconnection_template`.
PR 5998 will add the entries to settings.yml, once it is approved.

File: modules/apps_api/app/workers/apps_api/fetch_connections.rb
Comment (bradenshipley) [reply modules/apps_api/app/workers/apps_api/fetch_connections.rb]: Based on Saman's response here, I've just updated the Settings.yml and test.yml as well. Latest commit has the `Settings.vanotify.services.lighthouse.template_id.disconnection_template` naming convention added in his PR (https://github.com/department-of-veterans-affairs/vets-api/pull/5998/files/767be6e83acb9d3b3d3fe318b7c26ee597f9140f)

Comment (saman-moshafi-va) [reply modules/apps_api/app/workers/apps_api/fetch_connections.rb]: [Waiting for the PR 5998 approval](https://github.com/department-of-veterans-affairs/vets-api/pull/5998).

Comment (bradenshipley) [reply modules/apps_api/app/workers/apps_api/fetch_connections.rb]: I'll leave the changes to settings.yml up to your PR then

File: modules/apps_api/app/workers/apps_api/fetch_connections.rb
Comment (bradenshipley) [reply modules/apps_api/app/workers/apps_api/fetch_connections.rb]: Inside devops the vanotify block has `services`, but you're saying it's not going to?
```
vanotify:
  client_url: ...
  services:
    lighthouse:
      template_id:
        disconnection_template: ...
        connection_template: ...
```

So the template entry should be `Settings.vanotify.lighthouse.template_id.connection_template`, right?

Comment (saman-moshafi-va) [reply modules/apps_api/app/workers/apps_api/fetch_connections.rb]: Correct, the entry should be Settings.vanotify.services.lighthouse.template_id.connection_template.  Devops is already changed to set the values for review, dev, staging and prod environments.

Comment (bradenshipley) [reply modules/apps_api/app/workers/apps_api/fetch_connections.rb]: Ok, so the references the the template_ids are correct and I'll mark this part resolved.

**Review by charleystran — APPROVED** [review]
LGTM

**Pull request merged**
